### ISSUE TITLE
False Positive in 920220

### ISSUE THREAD

**Issue by theMiddleBlue** [coreruleset/coreruleset#1782]
### Description

920220 execute validateUrlEncoding over the whole REQUEST_URI. This lead to a false positive when a request comes with an argument value that contains a percentage (example `with=100%&heigth=100%`). Even if the argument `width` is correctly encoded (`width=100%25&...`) the rule triggers maybe because `%25&h` is not a valid hex sequence.

What about replacing REQUEST_URI with ARGS? Moreover, in the rule logic comment, we mention something about REQUEST_BODY. Is it related to something that has been removed?

### Audit Logs / Triggered Rule Numbers

Request:
`/admin/index.php?width=100%25&height=270`

Rule:
`920220 URL Encoding Abuse Attack Attempt`

https://github.com/coreruleset/coreruleset/blob/de26f0df363733ae3414ef20653ca8d5281188da/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf#L347-L379

**Comment by fgsch** [comment]
This is odd. %25 is valid and should not cause any issues.
Do you have any more details? Is this modsec 3 or 2?

**Comment by fgsch** [comment]
btw, the REQUEST_BODY comment is about the rule below, 920240 which is a sibling of 920220.

**Comment by theMiddleBlue** [comment]
Nginx + libmodsecurity 3.0.3
do you think could be related to nginx + v3 only?

I managed to reproduce it on my test env:

### test 1 (no encoding)
Request: `curl -v 'http://wordpress/test.php?w=100%&h=100'`
```
2020/05/31 12:29:47 [info] 10#0: *7 ModSecurity: Warning. Matched "Operator `ValidateUrlEncoding' with parameter `' against variable `REQUEST_URI' (Value: `/t
est.php?w=100%&h=100' ) [file "/opt/owasp-modsecurity-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "328"] [id "920220"] [rev ""] [msg "URL Encoding
 Abuse Attack Attempt"] [data "/test.php?w=100%&h=100"] [severity "4"] [ver "OWASP_CRS/3.2.0"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "l
anguage-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "OWASP_CRS/PROTOCOL_VIOLATION/EVASION"] [hostna
me "172.18.0.5"] [uri "/test.php"] [unique_id "159092098735.850103"] [ref "o15,1v4,22o15,22v4,22"], client: 172.18.0.1, server: wordpress, request: "GET /test
.php?w=100%&h=100 HTTP/1.1", host: "wordpress"
```
### test 2 (encoded % 0x25)
Request: `curl -v 'http://wordpress/test.php?w=100%25&h=100'`
```
2020/05/31 12:30:43 [info] 10#0: *10 ModSecurity: Warning. Matched "Operator `ValidateUrlEncoding' with parameter `' against variable `REQUEST_URI' (Value: `/
test.php?w=100%&h=100' ) [file "/opt/owasp-modsecurity-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "328"] [id "920220"] [rev ""] [msg "URL Encodin
g Abuse Attack Attempt"] [data "/test.php?w=100%&h=100"] [severity "4"] [ver "OWASP_CRS/3.2.0"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "
language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "OWASP_CRS/PROTOCOL_VIOLATION/EVASION"] [hostn
ame "172.18.0.5"] [uri "/test.php"] [unique_id "15909210438.071766"] [ref "o15,1v4,24o15,22v4,24"], client: 172.18.0.1, server: wordpress, request: "GET /test
.php?w=100%25&h=100 HTTP/1.1", host: "wordpress"
```

**Comment by fgsch** [comment]
IIRC when using REQUEST_URI the query string is not automatically url-decoded, so the rule should be seeing `/test.php?w=100%25&h=100`. I will confirm tomorrow but this sounds like a non-backward compatible change in modsec 3.

**Comment by theMiddleBlue** [comment]
Maybe we should start to put a condition to certain rules or to write a different version of the same rule to keep compatibility with v2 and v3. something like:

```
SecRule MODSEC_BUILD "@ge 030000000" "skipAfter: 920220"
... rule 920220 will work only on v2 ...

SecRule MODSEC_BUILD "@lt 030000000" "skipAfter: 92022x"
... new rule 92022x (sibling 920220) that works only on v3 ...
```

what do you think about it?

**Comment by fgsch** [comment]
> Maybe we should start to put a condition to certain rules or to write a different version of the same rule to keep compatibility with v2 and v3. something like:

Sounds like a topic for tonight's meeting 😄

**Comment by fgsch** [comment]
ModSecurity 2.9.3:

```
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Recipe: Invoking rule 7fa1d51ffc88; [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "352"] [id "920220"].
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][5] Rule 7fa1d51ffc88: SecRule "REQUEST_URI" "@rx \\x25" "phase:2,log,auditlog,id:920220,block,t:none,msg:'URL Encoding Abuse Attack Attempt',logdata:%{MATCHED_VAR},tag:application-multi,tag:language-multi,tag:platform-multi,tag:attack-protocol,tag:paranoia-level/1,tag:OWASP_CRS,tag:OWASP_CRS/PROTOCOL_VIOLATION/EVASION,ver:OWASP_CRS/3.2.0,severity:WARNING,chain"
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Transformation completed in 10 usec.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Executing operator "rx" with param "\\x25" against REQUEST_URI.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][9] Target value: "/test.php?w=100%25"
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Operator completed in 11 usec.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Rule returned 1.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][9] Match -> mode NEXT_RULE.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Recipe: Invoking rule 7fa1d51f93a8; [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "354"].
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][5] Rule 7fa1d51f93a8: SecRule "REQUEST_URI" "@validateUrlEncoding " "setvar:tx.anomaly_score_pl1=+%{tx.warning_anomaly_score}"
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Transformation completed in 9 usec.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Executing operator "validateUrlEncoding" with param "" against REQUEST_URI.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][9] Target value: "/test.php?w=100%25"
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Operator completed in 10 usec.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][4] Rule returned 0.
[01/Jun/2020:12:52:20 +0000] [localhost/sid#7fa1e3152110][rid#7fa1d6b940a0][/test.php][9] No match, not chained -> mode NEXT_RULE.
```

**Comment by dune73** [comment]
During the [monthly chat in June](https://github.com/coreruleset/coreruleset/issues/1779#issuecomment-637040773), we decided to keep this open until the July meeting, so we can gather more information about this.

**Comment by cyberbeat** [comment]
What about this?

**Comment by fzipi** [comment]
> What about this?

Good question. Do you have a point here?

**Comment by fzipi** [comment]
@airween Do you think this is still an issue in libmodsecurity?

**Comment by airween** [comment]
> @airween Do you think this is still an issue in libmodsecurity?

Not just in libmodsecurity3 - mod_security2 follows this way too. I'm investigating the problem.

**Comment by airween** [comment]
Take a look at the [rule](https://github.com/coreruleset/coreruleset/blob/f68ce69e3ef015fd1127c9415bb093eb431cc8ec/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf#L377-L399):

```
SecRule REQUEST_URI_RAW "@rx \x25" \
    "id:920220,\
    phase:1,\
    block,\
    t:none,t:urlDecodeUni,\
    msg:'URL Encoding Abuse Attack Attempt',\
    logdata:'%{REQUEST_URI_RAW}',\
    tag:'application-multi',\
    tag:'language-multi',\
    tag:'platform-multi',\
    tag:'attack-protocol',\
    tag:'paranoia-level/1',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/255/153/267/72',\
    ver:'OWASP_CRS/4.8.0-dev',\
    severity:'CRITICAL',\
    chain"
    SecRule REQUEST_URI_RAW "@rx ^(.*)/(?:[^\?]+)?(\?.*)?$" \
        "capture,\
        chain"
        SecRule TX:1|TX:2 "@validateUrlEncoding" \
            "t:none,t:urlDecodeUni,\
            setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
```

This is a chained rule with 3 rules.

In first rule the target is the `REQUEST_URI_RAW`, which value is `/?width=100%25&height=270`.

Also in the first rule the transformation `t:urlDecodeUni` transforms this URI into `/?width=100%&height=270`, so the `%25` will be replaced by a single `%` character, then `@rx` looks for a `%` (`\x25`) character.

If it founds then the second rule in the chain splits the original `REQUEST_RAW_URI` into 3 parts:
* `TX:0` will hold the whole string (`REQUEST_URI_RAW`)
* `TX:1` will be an empty string
* `TX:2` will hold this string: `?width=100%25&height=270` which is the stripped version of original value (without the leading `/`).

The third rule will check the `TX:1` (empty string) and `TX:2` (stripped `REQUEST_URI_RAW`), **but there is also a `t:urlDecodeUni` transformation**, so the `@validateUrlEncoding` receives this **decoded** RAW URI: `?width=100%&height=270`, which is definitely invalid, because the next character of `%` is an `&` sign.

If we left the `t:urlDecodeUni` transformation from the third rule, then it works as we expect - but I know this is not a solution, because we want to check if the URI is double encoded...

A side note: there is a false information, because we put the `REQUEST_URI_RAW` variable into the log (`logdata:'%{REQUEST_URI_RAW}'`), not the value that we processed.

**Comment by ZarPeter** [comment]
@airween  Thanks for the analysis. Does that mean we should add an exclusion for our rulesets for this, until a resolution is available?
I am rather new to OWASP rulesets and writing exlcusions...

As of today using OWASP 4.10 this issue still persists and is causing our customers some trouble. They have japanese filenames and also "%" characters in there...and those get transmitted as part of the url.

**Comment by airween** [comment]
> Does that mean we should add an exclusion for our rulesets for this, until a resolution is available? I am rather new to OWASP rulesets and writing exlcusions...

No worries, yes, you have to create an exclusion. Usually we bind the exclusion for an URL, so in your case this would be enough:

```
SecRule REQUEST_URI "@beginsWith /your/uri" \
    "id:999001,\
    phase:1,\
    t:none,\
    pass,\
    nolog,\
    ctl:ruleRemoveById=920220"
```
Add this rule into file `REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf` and restart your WAF.

**Comment by ZarPeter** [comment]
Thank you very much! This reolved my issue! 

Keep up this great work! :)
